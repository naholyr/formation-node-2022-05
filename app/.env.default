PORT=3001

# redis://username:authpassword@127.0.0.1:6380/4?allowUsernameInURI=true

REDIS_URL=redis://localhost:6379
TOKEN_DURATION=86400
