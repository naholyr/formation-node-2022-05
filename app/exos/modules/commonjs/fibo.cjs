const fibo = (n) => (n === 0 ? 1 : n === 1 ? 1 : fibo(n - 1) + fibo(n - 2));

// Shorthand properties
module.exports = { fibo };
