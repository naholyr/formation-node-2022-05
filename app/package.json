{
  "name": "app",
  "version": "1.0.0",
  "description": "nothing to see here",
  "main": "dist/index.js",
  "scripts": {
    "test": "jest",
    "hello": "echo hello",
    "build": "tsc",
    "clean": "rimraf dist",
    "tsc-watch": "tsc --watch",
    "nodemon": "nodemon dist/index.js",
    "dev": "npm-run-all -l -p tsc-watch nodemon"
  },
  "author": "Nicolas Chambrier",
  "license": "ISC",
  "dependencies": {
    "express": "^4.18.1",
    "lodash": "^4.17.21"
  },
  "devDependencies": {
    "@types/express": "^4.17.13",
    "@types/jest": "^27.5.1",
    "@types/lodash": "^4.14.182",
    "@typescript-eslint/eslint-plugin": "^5.23.0",
    "@typescript-eslint/parser": "^5.23.0",
    "eslint": "^8.15.0",
    "jest": "^28.1.0",
    "nodemon": "^2.0.16",
    "npm-run-all": "^4.1.5",
    "prettier": "^2.6.2",
    "rimraf": "^3.0.2",
    "ts-jest": "^28.0.2",
    "typescript": "^4.6.4"
  }
}
