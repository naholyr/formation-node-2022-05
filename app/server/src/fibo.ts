export const fibo = (n: number): number => {
  if (n < 0 || isNaN(n)) throw new Error("Invalid number");
  return fiboSafe(n);
};

const fiboSafe = (n: number): number => {
  if (n === 0) return 0;
  if (n === 1) return 1;
  return fiboSafe(n - 1) + fiboSafe(n - 2);
};
