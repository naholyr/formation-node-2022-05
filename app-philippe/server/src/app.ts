import express from "express";
import { fibo } from "./fibo";
import cors from "cors";
import { postsRouter } from "./routes/posts";
import bodyParser from "body-parser";
export const app = express();

app.use(
  cors({
    origin: (origin, callback) => callback(null, origin ? [origin] : true),
    credentials: true,
  })
);

app.use(bodyParser.json());

app.get("/fibo", (req, res) => {
  res.send({ result: fibo(40) });
});

app.get("/fibo/:nb", (req, res) => {
  res.send({ result: fibo(Number(req.params.nb)) });
});

app.get("/hello", (req, res) => {
  res.send("hello world");
});

app.use("/posts", postsRouter);
