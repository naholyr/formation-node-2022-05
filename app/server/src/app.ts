import express from "express";
import { fibo } from "./fibo";
import type { Handler } from "express";
import cors from "cors";
import { postsRouter } from "./routes/posts";
import bodyParser from "body-parser";

export const app = express();

app.use(
  cors({
    origin: (origin, callback) => callback(null, origin ? [origin] : true),
    credentials: true,
  })
);

app.use(bodyParser.json()); // application/json
// bodyParser.text() → text/plain
// bodyParser.raw() → application/octet-stream
// bodyParser.urlencoded() → application/x-www-form-urlencoded

const log: Handler = (req, res, next) => {
  // eslint-disable-next-line no-console
  console.log(new Date(), req.method, req.url);
  // Call next or chain is broken!
  next();
};

app.get("/*", log);

// Simplest GET handler
app.get("/hello", (req, res) => {
  res.send("hello world");
});

// Route parameters
app.get("/fibo/:nb([0-9]+)", (req, res) => {
  const input = Number(req.params.nb);
  res.send({ input, result: fibo(input) });
});

// Posts REST API
app.use("/posts", postsRouter);
