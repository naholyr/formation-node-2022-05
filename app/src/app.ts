import express from "express";
import { fibo } from "./fibo";
import type { Handler } from "express";

export const app = express();

const log: Handler = (req, res, next) => {
  // eslint-disable-next-line no-console
  console.log(new Date(), req.method, req.url);
  // Call next or chain is broken!
  next();
};

app.get("/*", log);

// Simplest GET handler
app.get("/hello", (req, res) => {
  res.send("hello world");
});

// Route parameters
app.get("/fibo/:nb([0-9]+)", (req, res) => {
  res.send({ result: fibo(Number(req.params.nb)) });
});
