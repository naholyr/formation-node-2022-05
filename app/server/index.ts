import { app } from "./src/app";

app.listen(3001, () => {
  // eslint-disable-next-line no-console
  console.log("Server ready http://localhost:3001");
});
