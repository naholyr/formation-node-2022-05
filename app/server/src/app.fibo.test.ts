import request from "supertest";
import { app } from "./app";
import { fibo } from "./fibo";

jest.mock("./fibo");

// Fonction utilitaire pratique pour manipuler les méthodes mockés
// Sans ça "fibo.mockReturnValue(…)" n'est pas reconnu par TS et un simple "as any" nous prive de l'auto-complétion
const mock = <T extends jest.MockableFunction>(fn: T): jest.MockedFn<T> => {
  return fn as unknown as jest.MockedFn<T>;
};

describe("App / Fibo", () => {
  it("GET /fibo/xxx should call fibo(xxx)", async () => {
    await request(app).get("/fibo/50");
    expect(fibo).toHaveBeenCalledWith(50);
  });
  it("GET /fibo/xxx should return { result: fibo(xxx) }", async () => {
    mock(fibo).mockReturnValue(42);
    const response = await request(app).get("/fibo/50");
    expect(response.status).toBe(200);
    expect(response.body).toMatchObject({ result: 42 });
    expect(response.body).toEqual(expect.objectContaining({ result: 42 }));
    expect(response.body).toEqual({ input: 50, result: 42 });
  });
  it("GET /fibo/xxx should not change output", async () => {
    mock(fibo).mockReturnValue(42);
    const response = await request(app).get("/fibo/50");
    expect(response.status).toBe(200);
    expect(response.body).toMatchSnapshot();
  });
});
