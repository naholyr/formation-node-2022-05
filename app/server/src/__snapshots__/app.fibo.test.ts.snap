// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App / Fibo GET /fibo/xxx should not change output 1`] = `
Object {
  "input": 50,
  "result": 42,
}
`;
